diff --git a/sql-common/client.c b/sql-common/client.c
index a453a1a..9e35ed7 100644
--- a/sql-common/client.c
+++ b/sql-common/client.c
@@ -1535,6 +1535,43 @@ unpack_fields(MYSQL_DATA *data,MEM_ROOT *alloc,uint fields,
   DBUG_RETURN(result);
 }
 
+ulong utf8_or_unibyte_length(uchar *buf, ulong size)
+{
+  ulong length= 0;
+  uchar *end= buf+size;
+  int rem= 0; /* UTF-8 code point remaining bytes */
+  for (; buf != end ; buf++)
+  {
+    if (rem)
+    {
+      if ((*buf & 0xC0) == 0x80)
+        --rem;
+      else
+        return size;
+    }
+    else
+    {
+      length++;
+      if (*buf < 0x80)
+        rem= 0;
+      else if (*buf < 0xC0)
+        return size;
+      else if (*buf < 0xE0)
+        rem= 1;
+      else if (*buf < 0xF0)
+        rem= 2;
+      else if (*buf < 0xF8)
+        rem= 3;
+      else
+        return size;
+    }
+  }
+  if (rem)
+    return size;
+  else
+    return length;
+}
+
 /* Read all rows (fields or data) from server */
 
 MYSQL_DATA *cli_read_rows(MYSQL *mysql,MYSQL_FIELD *mysql_fields,
@@ -1609,6 +1646,7 @@ MYSQL_DATA *cli_read_rows(MYSQL *mysql,MYSQL_FIELD *mysql_fields,
 	cp+=len;
 	if (mysql_fields)
 	{
+	  len = utf8_or_unibyte_length(cp-len, len);
 	  if (mysql_fields[field].max_length < len)
 	    mysql_fields[field].max_length=len;
 	}
